vm/src/stdlib/bisect.rs

```diff
@@ -0,0 +1,182 @@
+        })
+    }
+
+    // Handles defaults for lo, hi.
+    //
+    //  - lo must be >= 0 with a default of 0.
+    //  - hi, while it could be negative, defaults to `0` and the same effect is achieved
+    //    (while loop isn't entered); this way we keep it a usize (and, if my understanding
+    //    is correct, issue 13496 is handled). Its default value is set to the length of the
+    //    input sequence.
+    #[inline]
+    fn as_usize(
+        lo: OptionalArg<PyObjectRef>,
+        hi: OptionalArg<PyObjectRef>,
+        seq_len: usize,
+        vm: &VirtualMachine,
+    ) -> PyResult<(usize, usize)> {
+        // We only deal with positives for lo, try_from can't fail.
+        // Default is always a Some so we can safely unwrap.
+        let lo = handle_default(lo, Some(0), vm)?
+            .map(|value| {
+                if value < 0 {
+                    return Err(vm.new_value_error("lo must be non-negative".to_owned()));
+                }
+                Ok(usize::try_from(value).unwrap())
+            })
+            .unwrap()?;
+        let hi = handle_default(hi, None, vm)?
+            .map(|value| {
+                if value < 0 {
+                    0
+                } else {
+                    // Can't fail.
+                    usize::try_from(value).unwrap()
+                }
+            })
+            .unwrap_or(seq_len);
+        Ok((lo, hi))
+    }
+
+    #[inline]
+    fn bisect_left_impl(
+        BisectArgs { a, x, lo, hi }: BisectArgs,
+        vm: &VirtualMachine,
+    ) -> PyResult<usize> {
+        let (mut lo, mut hi) = as_usize(lo, hi, vm.obj_len(&a)?, vm)?;
+
+        while lo < hi {
+            // Handles issue 13496.
+            let mid = (lo + hi) / 2;
+            if vm.bool_cmp(&a.get_item(mid, vm)?, &x, Lt)? {
+                lo = mid + 1;
+            } else {
+                hi = mid;
+            }
+        }
+        Ok(lo)
+    }
+
+    #[inline]
+    fn bisect_right_impl(
+        BisectArgs { a, x, lo, hi }: BisectArgs,
+        vm: &VirtualMachine,
+    ) -> PyResult<usize> {
+        let (mut lo, mut hi) = as_usize(lo, hi, vm.obj_len(&a)?, vm)?;
+
+        while lo < hi {
+            // Handles issue 13496.
+            let mid = (lo + hi) / 2;
+            if vm.bool_cmp(&x, &a.get_item(mid, vm)?, Lt)? {
+                hi = mid;
+            } else {
+                lo = mid + 1;
+            }
+        }
+        Ok(lo)
+    }
+
+    /// Return the index where to insert item x in list a, assuming a is sorted.
+    ///
+    /// The return value i is such that all e in a[:i] have e < x, and all e in
+    /// a[i:] have e >= x.  So if x already appears in the list, a.insert(x) will
+    /// insert just before the leftmost x already there.
+    ///
+    /// Optional args lo (default 0) and hi (default len(a)) bound the
+    /// slice of a to be searched.
+    #[pyfunction]
+    fn bisect_left(args: BisectArgs, vm: &VirtualMachine) -> PyResult<usize> {
```

Comment:
You could have `bisect_left_impl` just *be* the function with the `#[pyfunction]` - you can still call functions with `#[pyfunction]` from other Rust functions.

Reply:
@coolreader18 The only thing I was unsure of is how `#[inline]` affects a `#[pyfunction]`, that's why I separated them (expecting a reviewer to point things out :)). Does it make no difference? If so, your suggestion is clearly the best approach.

Reply:
The `fn` item the macro's applied to *should* just be passed straight through (last time I looked at the macro code at least), so `#[inline]` and everything works like a normal function would.